src/permission/fs_permission.cc

```diff
@@ -0,0 +1,217 @@
+#include "fs_permission.h"
+#include "base_object-inl.h"
+#include "util.h"
+#include "v8.h"
+
+#include <fcntl.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <algorithm>
+#include <filesystem>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace {
+
+std::string WildcardIfDir(const std::string& res) noexcept {
+  uv_fs_t req;
+  int rc = uv_fs_stat(nullptr, &req, res.c_str(), nullptr);
+  if (rc == 0) {
+    const uv_stat_t* const s = static_cast<const uv_stat_t*>(req.ptr);
+    if (s->st_mode & S_IFDIR) {
+      // add wildcard when directory
+      if (res.back() == node::kPathSeparator) {
+        return res + "*";
+      }
+      return res + node::kPathSeparator + "*";
+    }
+  }
+  uv_fs_req_cleanup(&req);
+  return res;
+}
+
+void FreeRecursivelyNode(
+    node::permission::FSPermission::RadixTree::Node* node) {
+  if (node == nullptr) {
+    return;
+  }
+
+  if (node->children.size()) {
+    for (auto& c : node->children) {
+      FreeRecursivelyNode(c.second);
+    }
+  }
+
+  if (node->wildcard_child != nullptr) {
+    delete node->wildcard_child;
+  }
+  delete node;
+}
+
+bool is_tree_granted(
+    node::permission::FSPermission::RadixTree* deny_tree,
+    node::permission::FSPermission::RadixTree* granted_tree,
+    const std::string& param) {
+#ifdef _WIN32
+  // is UNC file path
+  if (param.rfind("\\\\", 0) == 0) {
```

Comment:
We call `path.toNamespacePath()` on `fs.js`. So, when using Windows a UNC file path might be passed to the `is_granted` call. Example: `\\?\C:\my_dir`.

The lookup uses a prefix radix tree algorithm, so I need to normalize the file path to `C:\my_dir` and I didn't find many resources on the web about converting UNC to absolute paths. Therefore, I just did the reverse engineering from `path.toNamespacePath` https://github.com/nodejs/node/blob/main/lib/path.js#L632.

I would appreciate it if someone else carefully review it @nodejs/platform-windows